ext/standard/streamsfuncs.c

```diff
@@ -756,115 +795,144 @@ static int stream_array_emulate_read_fd_set(zval *stream_array)
 /* }}} */
 
 /* {{{ Runs the select() system call on the sets of streams with a timeout specified by tv_sec and tv_usec */
-PHP_FUNCTION(stream_select)
-{
-	zval *r_array, *w_array, *e_array;
-	struct timeval tv, *tv_p = NULL;
-	fd_set rfds, wfds, efds;
-	php_socket_t max_fd = 0;
-	int retval, sets = 0;
-	zend_long sec, usec = 0;
-	bool secnull;
-	bool usecnull = 1;
-	int set_count, max_set_count = 0;
-
-	ZEND_PARSE_PARAMETERS_START(4, 5)
-		Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
-		Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
-		Z_PARAM_LONG_OR_NULL(sec, secnull)
-		Z_PARAM_OPTIONAL
-		Z_PARAM_LONG_OR_NULL(usec, usecnull)
-	ZEND_PARSE_PARAMETERS_END();
+PHP_FUNCTION(stream_select) {
+    zval *r_array, *w_array, *e_array;
+    struct timeval tv, *tv_p = NULL;
+    fd_bigset rfds, wfds, efds;
+    php_socket_t max_fd = 0;
+    int retval, sets = 0;
+    zend_long sec, usec = 0;
+    bool secnull;
+    bool usecnull = 1;
+    int set_count = 0, max_set_count = 0;
+
+    ZEND_PARSE_PARAMETERS_START(4, 5)
+        Z_PARAM_ARRAY_EX2(r_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(w_array, 1, 1, 0)
+        Z_PARAM_ARRAY_EX2(e_array, 1, 1, 0)
+        Z_PARAM_LONG_OR_NULL(sec, secnull)
+        Z_PARAM_OPTIONAL
+        Z_PARAM_LONG_OR_NULL(usec, usecnull)
+    ZEND_PARSE_PARAMETERS_END();
 
-	FD_ZERO(&rfds);
-	FD_ZERO(&wfds);
-	FD_ZERO(&efds);
+#ifdef PHP_WIN32
+	/* Check if the Windows standard handle is redirected to file */
+	long max_fds = FD_SETSIZE;
+#else
+	/* Check if the file descriptor identifier is a terminal */
+	long max_fds = sysconf(_SC_OPEN_MAX);
+#endif
 
-	if (r_array != NULL) {
-		set_count = stream_array_to_fd_set(r_array, &rfds, &max_fd);
+    FD_BIGSET_ZERO(&rfds, max_fds);
+    FD_BIGSET_ZERO(&wfds, max_fds);
+    FD_BIGSET_ZERO(&efds, max_fds);
```

Comment:
We should verify that `max_fds` is a valid, positive number. If not, it should probably default to `1024` (or whatever constant) to work like it used to.

Reply:
Can sysconf return a negative value for it? It does fall back to FD_SETSIZE if sysconf fails now.

Reply:
See [man sysconf](https://www.man7.org/linux/man-pages/man3/sysconf.3.html):

> The return value of sysconf() is one of the following:
> 
> On error, -1 is returned and [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the
          error (for example, EINVAL, indicating that name is invalid).
>
> If name corresponds to a maximum or minimum limit, and that
          limit is indeterminate, -1 is returned and [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) is not
          changed.  (To distinguish an indeterminate limit from an
          error, set [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) to zero before the call, and then check
          whether [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) is nonzero when -1 is returned.)
>
> If name corresponds to an option, a positive value is returned
          if the option is supported, and -1 is returned if the option
          is not supported.
>
> Otherwise, the current value of the option or limit is
          returned.  This value will not be more restrictive than the
          corresponding value that was described to the application in
          <unistd.h> or <limits.h> when the application was compiled.

So, theoretically, if `uname -n unlimited` were possible, you could also receive a `-1` too. AFAIK, it isn't possible to set the open file limit to unlimited though. (I could be wrong, but I'm 95% sure).

Reply:
The latest commit handles max_fds == -1 as well as max_fds > 131072 capping it. In addition, the bitmap size is doubled dynamically on demand by FD_BIGSET_SET.